### ISSUE TITLE
Previous median_filter functionality and current float [-1,1] requirement

### ISSUE THREAD

**Issue by dshean** [scikit-image/scikit-image#1616]
I just upgraded to 0.11.3 and I'm kind of surprised by the changes in the median filter functionality. I used the older skimage.filter.median_filter rather than the scipy.ndimage.filters.median_filter because I preferred the way it handled missing values. Scipy propagates data gaps (http://docs.astropy.org/en/v0.3/convolution/index.html), while the old skimage median filter would "fill" small gaps. It looks like the old skimage median_filter functionality is no longer available?

My primary inputs are large, single-band float32 or float64 numpy arrays (that often have data gaps).  I initially avoided upgrading skimage because of the lack of support for float arrays.  It looks like this has received some recent attention (https://github.com/scikit-image/scikit-image/issues/1419, https://github.com/scikit-image/scikit-image/issues/1529, https://github.com/scikit-image/scikit-image/pull/1422).  I tried to update my tools with the new function, but then received the `ValueError: Images of type float must be between -1 and 1.`  

It seems that requiring the user to scale their input float array to [-1,1] will be cumbersome. Hopefully this will happen behind the scenes in future releases.

**Comment by stefanv** [comment]
We don't aim to change any behaviour unless explicitly documented in api_changes.txt, so this was probably an oversight.

The pattern we have for handling images outside of the standard range is a "preserve_range" keyword--not sure which function you are using, it may not be implemented there.

@scikit-image/core Should we add preserve_range as a keyword to img_as_float and friends so that the pattern can be implemented across the package more easily?

**Comment by ahojnnes** [comment]
@stefanv preserve_range sounds like a good change to me.

**Comment by JDWarner** [comment]
We could easily decorate most of the package with a motif like this, exporting all of the range enforcement (untested):

``` python
# function enforce_dtype_ranges() conducts range enforcement
from functools import wraps

def preserve_range(func):
    @wraps(func)  # Preserve docstring/signature
    def with_preserve_range(*args, **kwargs):
        # Pop out additional kwarg if passed
        if 'preserve_range' in kwargs:
            preserve_range = kwargs.pop('preserve_range')
        else:
            preserve_range = False  # Default, could be set by decorator per-function

        # Enforce dtype limits if needed
        if preserve_range:
            return func(*args, **kwargs)
        else:
            return enforce_dtype_range(func(*args, **kwargs))
    return with_preserve_range
```

**Comment by blink1073** [comment]
:+1: for name and using a decorator

**Comment by stefanv** [comment]
Looks good to me! Josh, will you file a PR with that decorator + a test?

**Comment by ahojnnes** [comment]
How is enforce_dtype_range supposed to work?

**Comment by JDWarner** [comment]
@ahojnnes that's the trickier one, I started with the decorator for a reason ;)

Problem cases:
* dtypes "hiding" in larger array types (12 or 14bit raw camera images in a np.uint16 image)
* floating point images outside [-1, 1]

I see the first as not feasible to address in a satisfactory manner. We just leave those images alone. `preserve_range`, where used in the package presently, addresses the second point - if returned floating point images should be returned on the range [-1, 1]. Non-floating point dtypes "self-police" their allowed ranges by simply aliasing on assigned over- or under-range values.

So as imagined, I was going to write a function to check gracefully if output(s) had a dtype.kind = 'f', if so, `enforce_dtype_range` would clip the output to [-1, 1] and return. This would ideally be done for all outputs, but with a little extra complexity the decorator could accept an argument specifying which outputs are to be inspected.

**Comment by ahojnnes** [comment]
@JDWarner that is why I was asking. Don't you think this function would be
failing in way too many cases in unexpected or hard to understand ways?

**Comment by JDWarner** [comment]
I should think something like this would be rather robust, but comments/improvements welcome. (Untested, it's late, but you can see what I'm thinking)

``` python
def _clip_float_to_range(arr, amin=-1, amax=1]):
    """
    Clips array-like object with .clip() attribute to the specified range.
    """
    try:
        return arr.clip(amin, amax)
    except AttributeError:
        # Misidentified array or error during clipping; return original value.
        return arr
    


def enforce_dtype_range(variable, amin=-1, amax=1):
    """
    Attempts to clip any floating point arrays to [-1, 1].
    
    Parameters
    ----------
    variable : Python object
        Output from any function.
    amin : float
        Minimum allowed value. Default: -1.
    amax : float
        Maximum allowed value. Default: 1.
    
    Returns
    -------
    variable : Python object
        Clipped array or arrays in function output.
    """
    # Shortcut trivial case
    if variable is None:
        return
    
    # Single numpy array output
    if isinstance(variable, np.ndarray):
        if variable.dtype.kind == 'f':
            return _clip_float_to_range(arr, amin=amin, amax=amax)

    # For a single tuple, suspect multiple return values.
    # Look one level inside.
    if isinstance(variable, tuple):
        # Attempt to clip return values individually
        return tuple(_clip_float_to_range(x, amin=amin, amax=amax) 
                     for x in variable)

    # Unforseen case, simply pass through.
    else:
        return variable
```

**Comment by ahojnnes** [comment]
Hm... it seems like you have a different expectation of what preserve_range
should do: rescaling the range of values vs. clipping. I don't think this
is what I would expect from "preserve_range=True" (emphasis on "I" as other
people might think differently of course) and what is currently implement
e.g. in the warping functions. I expect the following: if I pass in an
array of any type, the function does not mess with the range of values
simply because of data type conversion. E.g., there is no obvious reason
that a warping function rescales the range of values just because it has to
"up-cast" the values to float/double. However, this often happens right now
as our img_as_* functions rescale the range of values automatically. A
default value of "preserve_range=True" could solve this issue, even though
we need to be careful since some functions may actually rely on the
rescaling. Instead of implementing `preserve_range` on a package-wide using
a generic function, I think this must be done on a per-function basis to
apply this behavior to the appropriate input and output parameters.

**Comment by JDWarner** [comment]
@ahojnnes good points, I may have misunderstood the context re:preserve_range. it's easy enough to grab the input range and scale the output back instead of clipping.

The decorator could be applied judiciously, but it is a common enough motif that this approach would let us use a single common implementation.

A perhaps more definitive solution would be to refactor `img_as_*` commands from within functions to decorator wrappers for more uniform behavior.

**Comment by ahojnnes** [comment]
@JDWarner Scaling the output back to the input range only works for linear-functions though. I'll be more precise what I mean by preserve_range:
```
def function(image, preserve_range):
    image = img_as_float(image, preserve_range)
    return image**2

image = np.array([1], dtype=np.uint8)

function(image, preserve_range=True) -> 1**2
function(image, preserve_range=False) -> (1/255)**2

```

**Comment by jni** [comment]
I agree with @ahojnnes interpretation and that our default should be `preserve_range=True`. Our contract with our users should be something along the lines of, "we will not mess with your data unless absolutely necessary". This includes both clipping and scaling.

@JDWarner in addition, I don't think iterating over a tuple is a good plan at all. Very often, the additional returned arrays of a function are not images at all but things like transformations, indexes, etc. Clipping in these scenarios would be disastrous.

**Comment by JDWarner** [comment]
Simple enough to tell it which output(s) are going to need looked at, since we'd be wrapping the function manually.

**Comment by grlee77** [comment]
`skimage.filters.median` currently preserves range. In #6319, we are moving to more uniformly preserve range across the library
